Source/JavaScriptCore/jit/JITArithmetic.cpp

```diff
@@ -157,66 +198,84 @@ void JIT::emit_op_unsigned(const JSInstruction* currentInstruction)
+    };
+    emit_compareImpl(op1, op2, condition, emitCompareAndJump);
 }
 
-void JIT::emit_compareAndJumpImpl(VirtualRegister op1, VirtualRegister op2, unsigned target, RelationalCondition condition)
+template <typename EmitCompareFunctor>
+ALWAYS_INLINE void JIT::emit_compareImpl(VirtualRegister op1, VirtualRegister op2, RelationalCondition condition, const EmitCompareFunctor& emitCompare)
 {
     // We generate inline code for the following cases in the fast path:
     // - int immediate to constant int immediate
     // - constant int immediate to int immediate
     // - int immediate to int immediate
 
-    bool disallowAllocation = false;
-    if (isOperandConstantChar(op1)) {
-        emitGetVirtualRegister(op2, jsRegT10);
-        addSlowCase(branchIfNotCell(jsRegT10));
+    auto handleConstantCharOperand = [&](VirtualRegister left, VirtualRegister right, JSValueRegs rightJSReg, RelationalCondition cond) {
+        if (!isOperandConstantChar(left))
+            return false;
+
+        emitGetVirtualRegister(right, rightJSReg);
+        addSlowCase(branchIfNotCell(rightJSReg));
         JumpList failures;
-        emitLoadCharacterString(jsRegT10.payloadGPR(), regT0, failures);
+        emitLoadCharacterString(rightJSReg.payloadGPR(), rightJSReg.payloadGPR(), failures);
         addSlowCase(failures);
-        addJump(branch32(commute(condition), regT0, Imm32(asString(getConstantOperand(op1))->tryGetValue(disallowAllocation)[0])), target);
+        emitCompare(commute(cond), rightJSReg, Imm32(asString(getConstantOperand(left))->tryGetValue(false)[0]));
+        return true;
+    };
+
+    if (handleConstantCharOperand(op1, op2, jsRegT10, condition))
         return;
-    }
-    if (isOperandConstantChar(op2)) {
-        emitGetVirtualRegister(op1, jsRegT10);
-        addSlowCase(branchIfNotCell(jsRegT10));
-        JumpList failures;
-        emitLoadCharacterString(jsRegT10.payloadGPR(), regT0, failures);
-        addSlowCase(failures);
-        addJump(branch32(condition, regT0, Imm32(asString(getConstantOperand(op2))->tryGetValue(disallowAllocation)[0])), target);
+    if (handleConstantCharOperand(op2, op1, jsRegT10, commute(condition)))
         return;
-    }
-    if (isOperandConstantInt(op2)) {
-        emitGetVirtualRegister(op1, jsRegT10);
-        emitJumpSlowCaseIfNotInt(jsRegT10);
-        int32_t op2imm = getOperandConstantInt(op2);
-        addJump(branch32(condition, jsRegT10.payloadGPR(), Imm32(op2imm)), target);
+
+    auto handleConstantIntOperand = [&](VirtualRegister left, VirtualRegister right, JSValueRegs rightJSReg, RelationalCondition cond) {
+        if (!isOperandConstantInt(left))
+            return false;
+
+        emitGetVirtualRegister(right, rightJSReg);
+        emitJumpSlowCaseIfNotInt(rightJSReg);
+        emitCompare(commute(cond), rightJSReg, Imm32(getOperandConstantInt(left)));
+        return true;
+    };
+
+    if (handleConstantIntOperand(op1, op2, jsRegT32, condition))
```

Comment:
I think there is no reason to use jsRegT32 here. Let's use jsRegT10.

Reply:
I think there're reasons to use jsRegT32. Tried to replaced that with jsRegT10 but got some stress test failures.Need to figure out why.

Reply:
OK, this is true, this needs to be `jsRegT32`. The reason is that, we are jumping to the slow path in L261. And in the slow path, we are assuming that the target value is in `jsRegT32` in this case. So it needs to be aligned.
Let's double check that our new code here is not breaking the assumption, and let's add comment about it here.